### PR TITLE
Small adjustments to metadata fields fetching

### PR DESCRIPTION
**Pull request by javanna**
While looking at #106325, which moves fetching of metadata fields out of the `StoredFieldsPhase`, I noticed some small adjustments that we can make to `FieldsVisitor`, `CustomFieldsVisitor` and `StoredFieldsPhase`.

These are not functional changes, the only goal is to make things simpler and clearer, hopefully.

- add test coverage for situation where `_routing` is provided with docs, hence returned by default
- make a stronger connection between `CustomFieldsVisitor` and `FieldsVisitor` around fields that are treated differently (`_ignored`, `_routing`, `_id` and `_source`)
- explicitly exclude `_id` from `StoredFieldsPhase` like we already do for `_source` as it's retrieved separately
- move the `_source` exclusion in `StoredFieldsPhase` to after calling `getMatchingFieldNames`, so that patterns that match `_source` still exclude it

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/fieldvisitor/FieldsVisitor.java [hunk 1]
```diff
@@ -147,12 +147,6 @@ public void doubleField(FieldInfo fieldInfo, double value) {
         addValue(fieldInfo.name, value);
     }
 
-    public void objectField(FieldInfo fieldInfo, Object object) {
-        assert IdFieldMapper.NAME.equals(fieldInfo.name) == false : "_id field must go through binaryField";
-        assert sourceFieldName.equals(fieldInfo.name) == false : "source field must go through binaryField";
-        addValue(fieldInfo.name, object);
-    }
-
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/fieldvisitor/FieldsVisitor.java:155]: this was unused, never called

File: server/src/main/java/org/elasticsearch/index/fieldvisitor/FieldsVisitor.java [hunk 2]
```diff
@@ -178,7 +172,9 @@ public Map<String, List<Object>> fields() {
     }
 
     public void reset() {
-        if (fieldsValues != null) fieldsValues.clear();
+        if (fieldsValues != null) {
+            fieldsValues.clear();
+        }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/fieldvisitor/FieldsVisitor.java:177]: just missing curly brackets here, nothing to see

File: server/src/main/java/org/elasticsearch/index/fieldvisitor/FieldsVisitor.java [hunk 3]
```diff
@@ -193,11 +189,7 @@ void addValue(String name, Object value) {
             fieldsValues = new HashMap<>();
         }
 
-        List<Object> values = fieldsValues.get(name);
-        if (values == null) {
-            values = new ArrayList<>(2);
-            fieldsValues.put(name, values);
-        }
+        List<Object> values = fieldsValues.computeIfAbsent(name, k -> new ArrayList<>(2));
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/fieldvisitor/FieldsVisitor.java:192]: shorter version but functionally the same.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/StoredFieldsPhase.java [hunk 4]
```diff
@@ -73,16 +74,19 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
...
-                        }
-                        storedFields.add(new StoredField(fieldName, ft, sec.isMetadataField(ft.name())));
-                        fieldsToLoad.add(ft.name());
+                Collection<String> fieldNames = sec.getMatchingFieldNames(field);
+                for (String fieldName : fieldNames) {
+                    // _id and _source are always retrieved anyway, no need to do it explicitly. See FieldsVisitor.
+                    // They are not returned as part of HitContext#loadedFields hence they are not added to documents by this sub-phase
+                    if (IdFieldMapper.NAME.equals(field) || SourceFieldMapper.NAME.equals(field)) {
+                        continue;
+                    }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/StoredFieldsPhase.java:83]: There are two changes here:

- filter _source out after calling getMatchingFieldNames. that was an oversight. Why would we filter it out when requested explicitly but not when a pattern expands to _source. In any case, _source has special treatment and requesting it explicitly has no effect
- in addition to filtering _source out, filter _id out too. We end up asking for it but it is removed from the list later anyways in CustomFieldsVisitor, as it is always loaded.

**Review by piergm — APPROVED** [review]
LGTM! 👍

**Comment by javanna** [comment]
thanks @piergm !

**Pull request merged**